net/ping/ping.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package ping allows sending ICMP echo requests to a host in order to
+// determine network latency.
+package ping
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"tailscale.com/net/netns"
+	"tailscale.com/types/logger"
+)
+
+type response struct {
+	t   time.Time
+	err error
+}
+
+// Pinger represents a set of ICMP echo requests to be sent at a single time.
+//
+// A new instance should be created for each concurrent set of ping requests;
+// this type should not be reused.
+type Pinger struct {
+	c       net.PacketConn
+	Logf    logger.Logf
+	Verbose bool
+	timeNow func() time.Time
+	id      uint16 // uint16 per RFC 792
+	wg      sync.WaitGroup
+
+	// Following fields protected by mu
+	mu    sync.Mutex
+	seq   uint16 // uint16 per RFC 792
+	pings map[uint16]chan response
+}
+
+// New creates a new Pinger. The Context provided will be used to create
+// network listeners, and to set an absolute deadline (if any) on the net.Conn
+func New(ctx context.Context, logf logger.Logf) (*Pinger, error) {
+	p, err := newUnstarted(ctx, logf)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start by setting the deadline from the context; note that this
+	// applies to all future I/O, so we only need to do it once.
+	deadline, ok := ctx.Deadline()
+	if ok {
+		if err := p.c.SetReadDeadline(deadline); err != nil {
+			return nil, err
+		}
+	}
+
+	p.wg.Add(1)
+	go p.run(ctx)
+	return p, nil
+}
+
+func newUnstarted(ctx context.Context, logf logger.Logf) (*Pinger, error) {
+	var id [2]byte
+	_, err := rand.Read(id[:])
```

Comment:
optional: it would be a bit better for testing purposes to make it possible to pass in a random number source rather than using the system random number generator. As is, it's easy enough for a test to just override `id` after, so it's not a big problem.